### PR TITLE
#411 preserve rootPath in router

### PR DESCRIPTION
**Pull request by Esardes**

### CODE REVIEW

**Comment by greenhost87** [comment]
Why Root path is `String` parameter, when you can use `Path` as parameter type for application root.
If change type you automatically get access to `PathAndQuery.++` and can remove ugly code `uri.mkString.drop(1)`

PlusPlus method already work with `Root` without any magic, take a look at `PathAndQuerySpec.".++" should "correct concatenate complex path"`

But for performance reasons I think need add  check `if (rootPath == Root) uri else rootPath ++ uri`

**Comment by Esardes** [comment]
I agree if `PathAndQuery.++` existed it would be the way to do it, but it doesn't, it's only part of `Path`.

Agreed `rootPath` is better being a `Path`, done. Ideally I think it should also be the case in `KorolevServiceConfig` but I don't want to make a breaking change to the "public interface" unless you think it's best.

**Comment by greenhost87** [comment]
You can add `Path.++(tail: PathAndQuery): PathAndQuery` with Spec and after use it

**Comment by greenhost87** [comment]
`mkString` have bad performance, plus `.drop(1)` create copy of string and than is parse again with `.fromString` also not very fast method

Performance info: https://github.com/fomkin/korolev/pull/392

**Comment by greenhost87** [comment]
@fomkin may be good time to change `KorolevServiceConfig.rootPath` from `String` to `Path` ?

**Comment by greenhost87** [comment]
@Esardes Also take a look at https://github.com/fomkin/korolev/blob/63733f3dcd01eda083aa963dfddbe32f3f48b91d/modules/korolev/src/main/scala/korolev/server/internal/KorolevServiceImpl.scala I think there also need provide some changes otherwise you application can't work correctly

**Comment by greenhost87** [comment]
LGTM

File: modules/korolev/src/main/scala/korolev/internal/ApplicationInstance.scala [hunk 1]
```diff
@@ -101,9 +102,9 @@ final class ApplicationInstance
     for {
       snapshot <- stateManager.snapshot
       // Set page url if router exists
-      _ <-  router.fromState
-          .lift(snapshot(Id.TopLevel).getOrElse(initialState))
-          .fold(Effect[F].unit)(uri => frontend.changePageUrl(uri))
+      _ <- router.fromState
+        .lift(snapshot(Id.TopLevel).getOrElse(initialState))
+        .fold(Effect[F].unit)(uri => frontend.changePageUrl(if (rootPath == Root) uri else rootPath / uri.mkString))
```
Comment (fomkin) [inline modules/korolev/src/main/scala/korolev/internal/ApplicationInstance.scala:107]: 1. You implemented `++` below. Why do don't use it here?
2. It's may work but it doesn't correct. For example if `uri == Root / "c" / d`, and `rootPath = Root / "a" / b` then `rootPath / uri.mkString == /(/(/(Root, "a"), "b"), "c/d")`. It is not valid path object.

Comment (Esardes) [reply modules/korolev/src/main/scala/korolev/internal/ApplicationInstance.scala]: You're right, done
Also needed to fix root path passed to js, let me know if the way I did it is too hacky (js assumes rootPath ends with `/`)

I looked briefly at `ZioHttpKorolev` and it seems to contains some logic about checking requests against the rootPath. the `akka` and `http4s` didn't seem to. Does it mean something more needs to be done in korolev core, e.g in `KorolevServiceImpl`?

File: interop/zio-http/src/main/scala/korolev/zio/http/ZioHttpKorolev.scala [hunk 2]
```diff
@@ -23,7 +23,7 @@ class ZioHttpKorolev[R] {
 
     val korolevServer = korolev.server.korolevService(config)
 
-    val rootPath = Path(config.rootPath)
+    val rootPath = Path(config.rootPath.mkString)
```
Comment (greenhost87) [inline interop/zio-http/src/main/scala/korolev/zio/http/ZioHttpKorolev.scala:26]: You already have `Path` type in `Config.rootPath` what reason to convert it to `String` and than back to `Path`

Comment (Esardes) [reply interop/zio-http/src/main/scala/korolev/zio/http/ZioHttpKorolev.scala:26]: agreed it's confusing, but it's another `Path` defined within `korolev.zio.http`

File: examples/http4s-zio/src/main/scala/Http4sZioExample.scala [hunk 3]
```diff
@@ -1,42 +1,43 @@
...
   type AppTask[A] = RIO[ZEnv, A]
 
   private class Service()(implicit runtime: Runtime[ZEnv])  {
 
-    import levsha.dsl._
-    import html._
-    import scala.concurrent.duration._
+    import scala.concurrent.duration.*
+
+    import levsha.dsl.*
```
[the remark this answers is not in the document]
Comment (Esardes) [reply examples/http4s-zio/src/main/scala/Http4sZioExample.scala]: I didn't test but then was not a change I made on purpose (auto format IDE), reverted to `._` imports

File: modules/korolev/src/main/scala/korolev/internal/ApplicationInstance.scala (hunk 1 above)
Comment (Esardes) [reply modules/korolev/src/main/scala/korolev/internal/ApplicationInstance.scala]: > I looked briefly at ZioHttpKorolev and it seems to contains some logic about checking requests against the rootPath. the akka and http4s didn't seem to. Does it mean something more needs to be done in korolev core, e.g in KorolevServiceImpl?

@fomkin if the answer to this question is "I don't think there is anything that needs to be done at the moment" then I think I have replied to all comments

**Pull request merged**